### ISSUE TITLE
Fragment identifiers with ':'

### ISSUE THREAD

**Issue by gkellogg** [w3c/json-ld-syntax#66]
A bug was noted where the following was not expanded properly:

```json
{
  "@context": {
    "@base": "https://ex.org/",
    "u": {"@id": "urn:u:", "@type": "@id"}
  },
  "u": ["#Test", "#Test:2"]
}
```

The issue is that the expand IRI algorithm splits "#Test:2" on the colon, where it is intended to be part of the fragment identifier, which is legal.

Either the algorithm needs to be made more complex by detecting this, or we should explicitly say that fragment identifiers containing ':' are not supported.

cc /@davidlehn

**Comment by azaroth42** [comment]
Proposal:  Accept the bug, errata 1.0 and fix in 1.1.  We should allow : as a valid character, as it is valid according to the URI RFC (afaicr!)

R

**Comment by ajs6f** [comment]
👍 . I can think of a few examples of `:`- containing fragments in the wild.

**Comment by gkellogg** [comment]
This will create a few corner cases that need to be addressed, for example, creating a term starting with a "#" could create issues:

```json
{
  "@context": {
    "@base": "https://ex.org/",
    "u": {"@id": "urn:u:", "@type": "@id"},
    "#Test": "http://ex.com/"
  },
  "u": ["#Test", "#Test:2"]
}
```

renders

```
[
  {
    "urn:u:": [
      {
        "@id": "https://ex.org/#Test"
      },
      {
        "@id": "http://ex.com/2"
      }
    ]
  }
]
```

We should probably restrict terms from beginning with a "#", or at least provide a warning.

Additionally, the expand IRI function will need some provision to consider if it is not a compact IRI (given no defined term), then value may be considered to be a relative IRI to the document base.

There are certainly more corner cases, and enumerating specific tests would be useful to ensure coverage.

What about "a:b", where base includes "#"? Would that now expand to a relative IRI?

**Comment by iherman** [comment]
Just for the records...

@azaroth42 said
> as it is valid according to the URI RFC (afaicr!)

I looked it up in [rfc3986](https://tools.ietf.org/html/rfc3986#section-3.5), which says:

```
fragment    = *( pchar / "/" / "?" )
...
pchar   = unreserved / pct-encoded / sub-delims / ":" / "@"
```

Ie, `:` is a valid character in a fragment ID, we should not disallow it.

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2018/2018-09-28-json-ld).

- `RESOLVED:  Accept the bug, submit errata 1.0 and fix in 1.1 by tightening the definition of a URI scheme`
<details><summary><i class="differentiate">View the transcript</i></summary>
4. New Issues<br/>
4.1. Fragment identifiers with “:”<br/>
Ivan Herman: link: <a href="https://github.com/w3c/json-ld-syntax/issues/66">https://github.com/w3c/json-ld-syntax/issues/66</a><br/>
Rob Sanderson: moving on to issues<br/>
… This was discovered as a bug over the past day<br/>
… We’ve been discussing it, and it’s valid to have : in fragment ids, some we need to fix this in the spec<br/>
Gregg Kellogg: I think that issue is that JSON-LD does not have a syntax for compact IRIs vs. URLs like RDFa so you can recognize prefixes vs. what is a URI scheme<br/>
… in this case, this is a bug<br/>
… since the pound sign is not a scheme, and it’s not a defined term<br/>
… if it were, there wouldn’t be a concern.<br/>
… the case we’re running across is that #test and #test:2 the first is a relative URI, since it’s interpreted as a ID<br/>
… so it’s a IRI or a document-relative IRI. The second case, it could be a compact IRI or an absolute IRI<br/>
… by tightening up the language, it must either match a term or the scheme production and a hash test would not match that production<br/>
… but there are other corner cases<br/>
… where there are other places where the hash sign may occur<br/>
… we need to come up with more breaking cases and tests<br/>
… what about a:b where the base includes a hash?<br/>
… a does match the production of the scheme, and could be an absolute IRI<br/>
Rob Sanderson: so long as it’s deterministic, but it’s these cases where it’s legal, but not permitted according to the algorithm where there’s the issue<br/>
… for fixing it, have we tried it?<br/>
Gregg Kellogg: I haven’t fixed it, but I think it’s straightforward if we tighten up what a URI scheme is<br/>
Rob Sanderson: We can fix this in 1.1, but we should also errata 1.0<br/>
Gregg Kellogg: I think we should not change the requirements, but highlight the issue<br/>
Ivan Herman: has it yet been submitted in the errata list?<br/>
Gregg Kellogg: no<br/>
Ivan Herman: We should document it<br/>
… editorially, when we have the list of changes, we should make a reference to the errata so we can change 1.0 because it’s a bug<br/>
Proposed resolution: Accept the bug, submit errata 1.0 and fix in 1.1 by tightening the definition of a URI scheme (Rob Sanderson)<br/>
Ivan Herman: this is explicit in our charter<br/>
Gregg Kellogg: +1<br/>
Adam Soroka: +1<br/>
Ivan Herman: +1<br/>
Rob Sanderson: +1<br/>
David Newbury: +1<br/>
Jeff Mixter: +1<br/>
David I. Lehn: +1<br/>
Simon Steyskal: +1<br/>
Resolution #3: Accept the bug, submit errata 1.0 and fix in 1.1 by tightening the definition of a URI scheme<br/>
Adam Soroka: Process: When we submit an errata, are we supposed to do something, or is it an action?<br/>
Ivan Herman: just describing it<br/>
</details>
